cuda_core/cuda/core/style-guide.md

```diff
@@ -0,0 +1,1818 @@
+# CUDA Core Style Guide
+
+This style guide defines conventions for Python and Cython code in `cuda/core/experimental`.
+
+**This project follows [PEP 8](https://peps.python.org/pep-0008/) as the base style guide.** The rules in this document highlight project-specific conventions and extensions beyond PEP 8, particularly for Cython code and the structure of this codebase.
+
+## Table of Contents
+
+1. [File Structure](#file-structure)
+2. [Package Layout](#package-layout)
+3. [Import Statements](#import-statements)
+4. [Class and Function Definitions](#class-and-function-definitions)
+5. [Naming Conventions](#naming-conventions)
+6. [Type Annotations and Declarations](#type-annotations-and-declarations)
+7. [Docstrings](#docstrings)
+8. [Errors and Warnings](#errors-and-warnings)
+9. [Memory Management](#memory-management)
+10. [Thread Safety and Concurrency](#thread-safety-and-concurrency)
+11. [Cython-Specific Features](#cython-specific-features)
+12. [Constants and Magic Numbers](#constants-and-magic-numbers)
+13. [Comments and Inline Documentation](#comments-and-inline-documentation)
+14. [Code Organization Within Files](#code-organization-within-files)
+15. [Performance Considerations](#performance-considerations)
+16. [API Design Principles](#api-design-principles)
+17. [CUDA-Specific Patterns](#cuda-specific-patterns)
+18. [Copyright and Licensing](#copyright-and-licensing)
+
+---
+
+## File Structure
+
+Files in `cuda/core/experimental` must follow a consistent structure. The ordering of elements within a file is as follows:
+
+### 1. SPDX Copyright Header
```

Comment:
I'm -1 on including things in this guide that can be automated, and I realize that results in a lack of completeness.

In fact, I'd rather our lint rule for this particular thing just add it to the top of the file when it fails.

I think this can be done for all file types, including pyx files.